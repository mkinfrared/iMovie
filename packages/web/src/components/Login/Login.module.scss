@import "src/styles/common";

.Login {
  margin: 0 auto;
  max-width: 480px;
}
