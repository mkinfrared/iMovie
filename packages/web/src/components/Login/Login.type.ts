import { Dispatch } from "redux";

export interface LoginProps {
  open: boolean;
  onClose: () => void;
  openSingUp: () => void;
  dispatch: Dispatch;
}
