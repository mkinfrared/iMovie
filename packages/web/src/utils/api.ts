import axios, { AxiosError } from "axios";
import { OptionsObject, SnackbarKey, SnackbarMessage } from "notistack";

import NotificationAction from "ui/NotificationAction";
import { BASE_URL } from "utils/secrets";

type ShowSnackbar = (
  message: SnackbarMessage,
  options?: OptionsObject
) => SnackbarKey;

const api = axios.create({
  baseURL: BASE_URL,
  withCredentials: true
});

const configureInterceptors = (showSnackbar: ShowSnackbar) => {
  api.interceptors.response.use(
    (config) => config,
    (error: AxiosError) => {
      if (error.message === "Network Error") {
        showSnackbar("Network Error", {
          variant: "error",
          action: NotificationAction
        });

        return;
      }

      return Promise.reject(error);
    }
  );
};

export default api;

export { configureInterceptors };
