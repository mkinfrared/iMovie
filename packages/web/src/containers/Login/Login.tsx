import React from "react";
import { useDispatch } from "react-redux";

import LoginFC from "components/Login";
import { LoginProps } from "components/Login/Login.type";

const Login = (props: Omit<LoginProps, "dispatch">) => {
  const dispatch = useDispatch();

  return <LoginFC dispatch={dispatch} {...props} />;
};

export default React.memo(Login);
