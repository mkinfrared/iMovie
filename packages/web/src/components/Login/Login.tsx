import { loginValidation } from "@imovie/common";
import Button from "@material-ui/core/Button";
import Dialog from "@material-ui/core/Dialog";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContent from "@material-ui/core/DialogContent";
import DialogContentText from "@material-ui/core/DialogContentText";
import DialogTitle from "@material-ui/core/DialogTitle";
import TextField from "@material-ui/core/TextField";
import React from "react";
import { useForm } from "react-hook-form";

import usePasswordToggle from "hooks/usePasswordToggle";
import { loginUser } from "store/reducers/user/reducer";
import Transition from "ui/Transition";
import api from "utils/api";

import css from "./Login.module.scss";
import { LoginProps } from "./Login.type";

interface FormData {
  username: string;
  password: string;
}

const Login = ({ onClose, open, openSingUp, dispatch }: LoginProps) => {
  const { passwordVisible, renderAdornment } = usePasswordToggle();

  const { handleSubmit, errors, register, setError, formState } = useForm<
    FormData
  >({
    validationSchema: loginValidation,
    mode: "onBlur"
  });

  const onSubmit = handleSubmit(async (data) => {
    try {
      const response = await api.post("/auth", data);
      const user = response.data;

      onClose();

      dispatch(loginUser(user));
    } catch (e) {
      const responseErrors = Object.entries(e.response.data).map((entry) => {
        const [key, value] = entry as [keyof FormData, string[]];

        return {
          type: "required",
          name: key,
          message: value[0]
        };
      });

      setError(responseErrors);
    }
  });

  return (
    <Dialog
      open={open}
      className={css.Login}
      TransitionComponent={Transition}
      onClose={onClose}
    >
      <DialogTitle>Login</DialogTitle>
      <DialogContent>
        <form onSubmit={onSubmit} className={css.form} data-testid="form">
          <TextField
            name="username"
            className={css.input}
            label="Username"
            variant="outlined"
            helperText={errors.username?.message || null}
            error={!!errors.username}
            inputRef={register}
            inputProps={{ "data-testid": "username" }}
            fullWidth
          />
          <TextField
            name="password"
            className={css.input}
            label="Password"
            variant="outlined"
            type={passwordVisible ? "text" : "password"}
            helperText={errors.password?.message}
            error={!!errors.password}
            InputProps={{ endAdornment: renderAdornment() }}
            inputProps={{ "data-testid": "password" }}
            inputRef={register}
            fullWidth
          />
        </form>
        <DialogContentText>
          Need an account? Consider{" "}
          <button className={css.textButton} onClick={openSingUp} type="button">
            sign up
          </button>
        </DialogContentText>
      </DialogContent>
      <DialogActions>
        <Button onClick={onClose} color="secondary">
          Cancel
        </Button>
        <Button
          data-testid="loginButton"
          onClick={onSubmit}
          color="primary"
          disabled={formState.isSubmitting}
        >
          Login
        </Button>
      </DialogActions>
    </Dialog>
  );
};

export default React.memo(Login);
