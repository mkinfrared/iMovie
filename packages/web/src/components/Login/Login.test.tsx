import { act, fireEvent, render } from "@testing-library/react";
// eslint-disable-next-line import/order
import React, { ReactElement } from "react";
import "mutationobserver-shim";

import api from "utils/api";

import Login from "./Login";

jest.mock("utils/api");

describe("<Login />", () => {
  const onClose = jest.fn();
  const open = true;
  const dispatch = jest.fn();
  const openSignUp = jest.fn();
  const apiMock = api as jest.Mocked<typeof api>;

  let Component: ReactElement;

  beforeEach(() => {
    Component = (
      <Login
        open={open}
        onClose={onClose}
        dispatch={dispatch}
        openSingUp={openSignUp}
      />
    );

    jest.resetAllMocks();
  });

  it("should be defined", () => {
    const { baseElement } = render(Component, {});

    expect(baseElement).toBeDefined();
  });

  it("should match a snapshot", () => {
    const { baseElement } = render(Component);

    expect(baseElement).toMatchSnapshot();
  });

  it("should submit the form", async () => {
    apiMock.post.mockReturnValueOnce(Promise.resolve({ data: 42 }));

    const { getByTestId } = render(Component);
    const username = getByTestId("username") as HTMLInputElement;
    const password = getByTestId("password") as HTMLInputElement;
    const loginButton = getByTestId("loginButton");

    await act(async () => {
      await fireEvent.input(username, { target: { value: "marklar" } });

      await fireEvent.input(password, { target: { value: "foobar" } });
    });

    expect(username.value).toBe("marklar");

    expect(password.value).toBe("foobar");

    await act(async () => {
      await fireEvent.click(loginButton);
    });

    expect(apiMock.post).toHaveBeenCalled();

    expect(onClose).toHaveBeenCalled();

    expect(dispatch).toHaveBeenCalled();
  });

  it("should set errors on form fields", async () => {
    const data = { username: ["foobar"] };

    const fieldValues = {
      username: "marklar",
      password: "foobar"
    };

    apiMock.post.mockRejectedValueOnce({
      response: { data }
    });

    const { getByTestId, findByText } = render(Component);
    const username = getByTestId("username") as HTMLInputElement;
    const password = getByTestId("password") as HTMLInputElement;
    const loginButton = getByTestId("loginButton");

    await act(async () => {
      await fireEvent.input(username, {
        target: { value: fieldValues.username }
      });

      await fireEvent.input(password, {
        target: { value: fieldValues.password }
      });
    });

    expect(username.value).toBe(fieldValues.username);

    expect(password.value).toBe(fieldValues.password);

    await act(async () => {
      await fireEvent.click(loginButton);
    });

    expect(apiMock.post).toHaveBeenCalled();

    expect(onClose).not.toHaveBeenCalled();

    expect(dispatch).not.toHaveBeenCalled();

    const result = await findByText(data.username[0]);

    expect(result).toBeInTheDocument();
  });
});
