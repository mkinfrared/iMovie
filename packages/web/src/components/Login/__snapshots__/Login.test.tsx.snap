// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> should match a snapshot 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root Login"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper"
      role="none presentation"
      style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiPaper-elevation24 MuiPaper-rounded"
        role="dialog"
      >
        <div
          class="MuiDialogTitle-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6"
          >
            Login
          </h2>
        </div>
        <div
          class="MuiDialogContent-root"
        >
          <form
            class="form"
            data-testid="form"
          >
            <div
              class="MuiFormControl-root MuiTextField-root input MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                data-shrink="false"
              >
                Username
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  data-testid="username"
                  name="username"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-347 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-349"
                  >
                    <span>
                      Username
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root input MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                data-shrink="false"
              >
                Password
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
                  data-testid="password"
                  name="password"
                  type="password"
                  value=""
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
                >
                  <button
                    aria-label="toggle password visibility"
                    class="MuiButtonBase-root MuiIconButton-root"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        data-testid="invisible"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78l3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </button>
                </div>
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-347 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-349"
                  >
                    <span>
                      Password
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </form>
          <p
            class="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            Need an account? Consider
             
            <button
              class="textButton"
              type="button"
            >
              sign up
            </button>
          </p>
        </div>
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Cancel
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
            data-testid="loginButton"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Login
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
